### ISSUE TITLE
Prefill of prompts that fit in one prefill_step_size is about 3x slower than the chunked path on Gemma 4 (KV-shared layers run at full length)

### ISSUE THREAD

**Issue by uwu-420** [Blaizzy/mlx-vlm#2155]
`generate_step` prefills a prompt in one call of `_step` when it is not longer than `prefill_step_size` (default 2048). For Gemma 4 that call runs all 35 layers on the full sequence, evaluates the full-width logits, and only then slices with `logits_to_keep` (the slice from #1750 sits after the layer stack). The chunk loop instead calls `language_model(..., n_to_process, logits_to_keep=1)` and evaluates only the cache states, so the 20 KV-shared layers (15 to 34 on E2B) run on one row. Since those layers do not write KV, the full-length pass through them is wasted work.

On an M3 Pro with mlx-community/gemma-4-e2b-it-4bit, one 1011-token prompt, 8 greedy tokens, median of 3:

| mlx-vlm | mlx | one-step path (`prefill_step_size` 2048) | chunk loop (`prefill_step_size` 512) |
|---|---|---|---|
| 0.6.13 | 0.32.2 | 1.16 s, 870 t/s | 0.36 s, 2800 t/s |
| 0.6.17 | 0.32.2 | 1.16 s, 870 t/s | 0.36 s, 2800 t/s |
| main (2026-09-04) | 0.32.2 | 0.35 s | 0.35 s |

A direct `language_model` call on the same prompt takes 1.11 s when the logits are evaluated and 0.33 s when only the cache states are evaluated (`logits_to_keep=1`), so the difference is the last-row handling, not the version. A per-layer trace puts the hidden state after layer 14 at 0.36 s and after layer 34 at 1.11 s: the shared layers are about 68 percent of the forward at this length.

In end-to-end runs over about a hundred prompts of roughly 1000 tokens, and again with one 1120-token image per prompt, `prefill_step_size` 512 on 0.6.13 gave byte-identical outputs and saved about 0.7 s (text) and 0.9 s (image) of wall time per prompt.

Main already avoids this for the common case: #2119 (merged 2026-08-31) routes every prompt longer than one token through the chunk loop, so main shows 0.35 s on both settings. Paths that still bypass the chunk loop pay the full cost: `chunked_prefill_policy` returning False (bidirectional vision attention with image tokens, non-MTP drafters), and any caller that prefills through `_step` directly. The one-token `_step` remainder is fine.

Proposal: in `Gemma4TextModel`, when `logits_to_keep == 1` (or a new `last_row_only` flag), slice the hidden state to the last row at `first_kv_shared_layer_idx` before the shared layers run. That gives every path the chunk-loop speed independent of `prefill_step_size`, including policy-disabled chunking and speculative prefill.

<details><summary>Probe script (about 120 lines, needs only a Gemma 4 pack)</summary>

```python
#!/usr/bin/env python3
"""Probe: why is mlx-vlm's single-step prefill 3x slower than its chunked prefill on the same prompt?

Times, on one pack and one prompt of about 1000 tokens: generate() with prefill_step_size 2048 (the prompt fits in one
step: the _step path), 1010 (one chunk of 1010 tokens through the chunk loop plus a one-token _step), and 512; then
direct calls of model.language_model on the full prompt with a fresh cache, with and without n_to_process, with and
without logits_to_keep, each evaluated with mx.eval. Run:
  python prefill_probe.py --pack models/mlx-community__gemma-4-e2b-it-4bit --tokens 1024
"""
import argparse
import json
import random
import sys
import time

FILLER = ["The station recorded {w} for most of the day.", "Visibility stayed at {n} kilometres.",
          "Temperatures moved between {a} and {b} degrees.", "The forecast for the next day expects {w}.",
          "Pressure was {p} hectopascals at midday.", "Wind speed reached {n} kilometres per hour in gusts."]
WEATHER = ["light rain", "clear skies", "a cold wind from the north", "fog until noon", "scattered clouds",
           "a warm afternoon", "snow on the hills", "thunderstorms in the evening", "a mild breeze", "high humidity"]


def neutral_text(words: int, seed: int = 7) -> str:
    rng = random.Random(seed)
    out = []
    while sum(len(s.split()) for s in out) < words:
        out.append(rng.choice(FILLER).format(w=rng.choice(WEATHER), n=rng.randint(2, 60), a=rng.randint(-5, 15),
                                             b=rng.randint(16, 35), p=rng.randint(980, 1040)))
    return " ".join(out)


def timed(fn, repeats=3):
    fn()  # warm-up
    ts = []
    for _ in range(repeats):
        t = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t)
    return round(sorted(ts)[len(ts) // 2], 3)


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--pack", required=True)
    ap.add_argument("--tokens", type=int, default=1024)
    ap.add_argument("--out", default="")
    a = ap.parse_args()
    import mlx.core as mx
    import mlx_vlm
    from mlx_vlm import apply_chat_template, generate, load
    from mlx_vlm.utils import load_config

    model, processor = load(a.pack)
    config = load_config(a.pack)
    tok = processor.tokenizer if hasattr(processor, "tokenizer") else processor
    ids = tok.encode(neutral_text(6000), add_special_tokens=False)
    body = tok.decode(ids[: a.tokens - 40]) + "\n\nWhich weather condition is named most often? Reply in one sentence."
    prompt = apply_chat_template(processor, config, body, num_images=0)
    n = len(tok.encode(prompt, add_special_tokens=False))
    results = {"mlx": mx.__version__, "mlx_vlm": mlx_vlm.__version__, "prompt_tokens": n}

    for chunk in (2048, n - 1, 512):
        r = None

        def run():
            nonlocal r
            r = generate(model, processor, prompt, max_tokens=8, temperature=0.0, verbose=False, prefill_step_size=chunk)

        wall = timed(run)
        results[f"generate_chunk_{chunk}"] = {"wall_s": wall, "prompt_tps": round(r.prompt_tps), "prompt_tokens": r.prompt_tokens}
        print(f"generate prefill_step_size={chunk}: wall {wall}s, prompt {r.prompt_tokens} tokens at {r.prompt_tps:.0f} t/s", flush=True)

    # direct language-model calls on the full prompt (what _step and the chunk loop do underneath)
    from mlx_vlm.generate import ar as _ar  # the same cache module generate_step uses

    input_ids = mx.array(tok.encode(prompt, add_special_tokens=False))[None]
    lm = model.language_model
    emb_out = model.get_input_embeddings(input_ids, None, mask=None)
    embed = emb_out.inputs_embeds
    extra = {k: v for k, v in emb_out.to_dict().items() if k != "inputs_embeds" and v is not None}
    mx.eval(embed)
    print("embedding extras passed to the language model:", sorted(extra), flush=True)

    L = int(input_ids.shape[1])

    def call(n_tokens, eval_what, **kw):
        # the cache generate_step uses: the language model's own make_cache (sliding and full layers)
        pc = _ar.cache.make_prompt_cache(lm)
        out = lm(input_ids[:, :n_tokens], inputs_embeds=embed[:, :n_tokens], cache=pc, **extra, **kw)
        if eval_what == "logits":
            mx.eval(out.logits)
        else:  # what the chunk loop evaluates
            mx.eval([c.state for c in pc])

    variants = {
        # name: (tokens, what is evaluated, kwargs)
        "full_eval_logits": (L, "logits", {}),
        "full_eval_logits_keep1": (L, "logits", {"logits_to_keep": 1}),
        "full_eval_cache_only": (L, "cache", {"logits_to_keep": 1}),
        "chunkloop_replica_L-1_cache_only": (L - 1, "cache", {"n_to_process": L - 1, "logits_to_keep": 1}),
        "chunkloop_replica_L-1_eval_logits": (L - 1, "logits", {"n_to_process": L - 1, "logits_to_keep": 1}),
    }
    for name, (n_tokens, eval_what, kw) in variants.items():
        try:
            wall = timed(lambda: call(n_tokens, eval_what, **kw))
            results[f"lm_call_{name}"] = wall
            print(f"language_model {n_tokens} tokens, {name}: {wall}s", flush=True)
        except Exception as e:  # noqa: BLE001
            results[f"lm_call_{name}"] = f"error: {e}"[:200]
            print(f"language_model {n_tokens} tokens, {name}: error {e}", flush=True)

    # per-layer trace: evaluate only the hidden state after layer k (capture_layer_ids fills hidden_states in order)
    n_layers = len(lm.model.layers)
    first_shared = getattr(lm.model, "first_kv_shared_layer_idx", n_layers)
    for k in sorted({first_shared - 1, n_layers - 1}):
        def call_capture(k=k):
            pc = _ar.cache.make_prompt_cache(lm)
            out = lm(input_ids, inputs_embeds=embed, cache=pc, capture_layer_ids=[k], **extra)
            mx.eval(out.hidden_states[0])
        try:
            wall = timed(call_capture)
            results[f"lm_hidden_after_layer_{k}"] = wall
            print(f"language_model {L} tokens, hidden state after layer {k} only: {wall}s "
                  f"(layers {first_shared} to {n_layers - 1} share KV)", flush=True)
        except Exception as e:  # noqa: BLE001
            results[f"lm_hidden_after_layer_{k}"] = f"error: {e}"[:200]
            print(f"hidden after layer {k}: error {e}", flush=True)
    if a.out:
        with open(a.out, "a") as fh:
            fh.write(json.dumps(results) + "\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
```
</details>
